src/core/baseFactory.ts

```diff
@@ -0,0 +1,45 @@
+import { ARGUMENTS_SEPARATOR } from "./constants";
+import { IGasLimitEstimator } from "./interfaces";
+import { Transaction } from "./transaction";
+
+interface Config {
+    minGasLimit: bigint;
+    gasLimitPerByte: bigint;
+}
+
+/**
+ * @internal
+ */
+export class BaseFactory {
+    private gasConfig: Config;
+    private gasLimitEstimator?: IGasLimitEstimator;
+
+    constructor(options: { config: Config; gasLimitEstimator?: IGasLimitEstimator }) {
+        this.gasConfig = options.config;
+        this.gasLimitEstimator = options.gasLimitEstimator;
+    }
+
+    protected setTransactionPayload(transaction: Transaction, dataParts: string[]): void {
+        const data = dataParts.join(ARGUMENTS_SEPARATOR);
+        transaction.data = Buffer.from(data);
+    }
+
+    /**
+     * Sets the gas limit for the transaction.
+     * @param gasLimit - Optional gas limit to set. This is the value provided by the user.
+     * @param configGasLimit - Optional gas limit from the configuration. This is computed internally based on some config values.
+     */
+    protected async setGasLimit(transaction: Transaction, gasLimit?: bigint, configGasLimit?: bigint): Promise<void> {
+        if (gasLimit) {
```

Comment:
Maybe we can refactor to have `if - then do & return`, instead of `if - else if ... ...` - just an opinion?